text/287-cli-deprecation-warnings.md

```diff
@@ -0,0 +1,280 @@
+
+### Why should we _not_ do this?
+
+I see four reasons why we might not want to implement this feature:
+
+1. Deprecated elements are already highlighted by virtually any editor/IDE.
+  In my experience, customers are diligent in moving away from deprecated APIs,
+  so this feature might not provide much value.
+2. We risk overloading customers with warnings,
+  training them to ignore our output if there's too much of it.
+  A warning must be read in order to be effective, after all.
+3. Depending on the exact solution chosen
+  (see discussion below),
+  we might not be able to distinguish from the user's code using a deprecated API,
+  and from our own libraries using deprecated APIs.
+  Which means a user might get warnings that they will be unable to get rid of
+  (because they're coming from code that they don't control).
+4. Depending on the exact solution chosen
+  (see discussion below),
+  there might be edge cases where something is deprecated,
+  but we won't be able to report a particular pattern of its usage.
+  Some examples include:
+    * Using deprecated types only in type declarations.
+    * Static constants (which includes enums).
+    * Deprecated elements declared in the code, but not actually executed during synthesis.
+
+### What changes are required to enable this change?
+
+There are two high-level components of this change:
+
+#### 1. Changes to JSII
+
+We will add a feature to JSII that adds the warning code to the emitted JavaScript for deprecated elements.
+It will be off by default, and will have to be activated explicitly to affect the emitted JavaScript code.
+
+#### 2. Using the changed JSII
+
+Once we have modified JSII and released a new version of it,
+we will need to use it in the CDK,
+and start compiling CDK with the new option turned on.
+
+We should also modify our CDK test infrastructure to run with the `DEPRECATED`
+environment variable set to `error` by default,
+to prevent our own code from using deprecated APIs which would cause warnings to be shown to users.
+We should come up with a nice API that allows individual tests that are explicitly checking deprecated API(s)
+as regression tests to make the value of that environment variable `quiet` easily.
+
+### Is this a breaking change?
+
+No.
+
+### What are the drawbacks of this solution?
+
+1. Requires changes to JSII.
+2. Does not allow distinguishing between the customer using deprecated APIs,
+  and the library code using deprecated APIs.
+  We can alleviate this problem by making the value of the `DEPRECATED` environment variable `error` by default in our tests,
+  which will be a forcing function for us to stop using deprecated APIs in the Construct Library.
+3. Adding emitting warnings for accessing deprecated static constants
+  (including enums) will require huge changes to the emitted JavaScript
+  (changing from a field to a `static` getter).
+4. We only report APIs actually hit during a particular execution of the code
```

Comment:
Good call out (worth mentioning that this is displayed In the IdE of most languages)

Reply:
Done.